src/process_quantiles.R

```diff
@@ -0,0 +1,168 @@
+# Processing code to create percentiles for groups of HRUs
+
+# This is preliminary work.
+
+# Creating a historic understanding of how "total storage"
+#   has varied through time for each HRU by calculating
+#   quantiles to establish very low, low, normal, high, 
+#   and very high thresholds of "total storage" for each
+#   HRU at every day of the year. Rather than only using
+#   values at every day of the year, we are using 5 values
+#   before and 5 values after to calculate the thresholds 
+#   for each day.
+
+# Total storage = pkwater_equiv + soil_moist_tot + hru_intcpstor +
+#                 hru_impervstor + gwres_stor + dprst_stor_hru
+
+# TO DO: 
+#   1. Use task_id to loop through groups of HRUs using slurm - haven't been able to get a test working
+#   2. Calc for probs = seq(0, 1, by = 0.05)
+#   3. Fix this to work with incomplete years & leap years
+#   4. Add -Inf and +Inf in `combine_hru_quantiles` right before saving.
+
+start <- Sys.time()
+library(ncdf4) 
+library(data.table)
+
+########## Functions ########## 
+task_id_to_hru_seq <- function(task_id, n_hrus_per_task = 1000) {
+  # Convert task id to HRU ids (increment by n_hrus_per_task)
+  hru_id_start <- ((task_id-1)*n_hrus_per_task + 1) # which hru to start with
+  hru_id_end <- hru_id_start + (n_hrus_per_task-1) # which hru to end with
+  
+  # If we are past the last column, cut the final col id off
+  n_hrus <- 109951
+  hru_id_end <- ifelse(hru_id_end > n_hrus,
+                       yes = n_hrus,
+                       no = hru_id_end)
+  
+  hru_seq <- hru_id_start:hru_id_end
+  return(hru_seq)
+}
+########## ////// ########## 
+
+# Identify task id from yeti environment & convert to HRU ids ----
+task_id <- as.numeric(Sys.getenv('SLURM_ARRAY_TASK_ID', 'NA'))
+hru_seq <- task_id_to_hru_seq(task_id)
+hru_start <- head(hru_seq, 1)
+hru_end <- tail(hru_seq, 1)
+quantile_fn <- sprintf("grouped_quantiles/total_storage_quantiles_%s_to_%s.csv", hru_start, hru_end)
+
+if(!file.exists(quantile_fn)) {
+
+  message(sprintf("Started this task at %s", Sys.time()))
+
+  vars <- c(
+    "soil_moist_tot",
+    "pkwater_equiv",
+    "hru_intcpstor", 
+    "hru_impervstor",
+    "gwres_stor", 
+    "dprst_stor_hru"
+  )
+  
+  vars_data <- c()
+  for(var in vars) {
+    
+    message(sprintf("Reading in NetCDF data for %s", var))
+    
+    # The file is NetCDF
+    fn <- sprintf("historical_%s_out.nc", var)
```

Comment:
Yeah this could definitely be a bottleneck, if dozens of jobs start at once you could quickly overwhelm the read capacity of the disks.  It may be worthwhile to add a step to first move these files into the local scratch disk space on each node, and read from there, so there is less competition between processes.  You can access the local scratch space via the `${LOCAL_SCRATCH}` environment variable.  This batch script and R script use it: https://github.com/USGS-R/pgml_temperature_prediction/blob/master/slurm/lake_optimize.slurm#L31
https://github.com/USGS-R/pgml_temperature_prediction/blob/master/R/lake_optimize.R#L31
This can also be a good solution if you are limited by write speed, not sure that would be a factor here though.  Note that anything left in the scratch space gets wiped out when your job ends, so if you write output there you need to move it back to lustre or cxfs at the end of the job.  

Another simpler option would be to limit the number of simultaneous processes, i.e. with `#SBATCH -t 1-200%5` — that limits it to 5 jobs at once.   That could be good enough if the processing doesn't take that long after the read is done.

Reply:
I think scratch space is shared between all cores on the same node, so if you have a bunch of tasks all running on the same node (since each is only using 1 core, you could theoretically have up to 20 tasks on the same node I think), I think you will need to make sure each copy of the inputs goes into a subdirectory, so the reads stay separate.

Reply:
Yet another option would be to chunk up the netcdf inputs into individual files for each HRU.  You might still run into I/O limits there though as well.

Reply:
Yeah, the chunking into individual files was how it was written originally, but then @aappling-usgs and I talked about speeding up and wanted to see if we could do less reading/writing with intermediate data. Maybe I will kick it off overnight on my computer just to see if it will work locally in series.